src/networking/payload/introductionrequestpayload.rs

```diff
@@ -35,36 +59,158 @@ struct IntroductionRequestPayload {
+        conntype.1,
+        false,
+        false,
+        false,
+        false,
+        false,
+        self.advice,
+      ))
+      .add_u16(self.identifier)
+      .add_raw_remaining(self.extra_bytes.clone());
+    res
   }
 
   fn unpack(packet: Packet) -> Self {
+    let mut packetiter = packet.iter();
+
+    let destination_ipaddress = packetiter.next_raw(4).unwrap();
+    let destination_port = packetiter.next_u16().unwrap();
+
+    let source_lan_address = packetiter.next_raw(4).unwrap();
+    let source_lan_port = packetiter.next_u16().unwrap();
+
+    let source_wan_address = packetiter.next_raw(4).unwrap();
+    let source_wan_port = packetiter.next_u16().unwrap();
+
+    // flags are cc00000a where cc is two bits signifying the connection type and a signifies the advice
+    let flags = packetiter.next_bits().unwrap();
+
+    let identifier = packetiter.next_u16().unwrap();
+
+    let extra_bytes = packetiter.next_raw_remaining().unwrap();
 
     IntroductionRequestPayload {
-      destination_address: IPAddress {
-        address: String::from("1.2.3.4"),
-        port: 6421,
-        version: IPVersion::IPV4,
+      destination_address: Address {
+        address: Ipv4Addr::new(
+          destination_ipaddress[0],
+          destination_ipaddress[1],
+          destination_ipaddress[2],
+          destination_ipaddress[3],
+        ),
+        port: destination_port,
       },
-      source_lan_address: IPAddress {
-        address: String::from("1.2.3.4"),
-        port: 6421,
-        version: IPVersion::IPV4,
+      source_lan_address: Address {
+        address: Ipv4Addr::new(
+          source_lan_address[0],
+          source_lan_address[1],
+          source_lan_address[2],
+          source_lan_address[3],
+        ),
+        port: source_lan_port,
       },
-      source_wan_address: IPAddress {
-        address: String::from("1.2.3.4"),
-        port: 6421,
-        version: IPVersion::IPV4,
+      source_wan_address: Address {
+        address: Ipv4Addr::new(
+          source_wan_address[0],
+          source_wan_address[1],
+          source_wan_address[2],
+          source_wan_address[3],
+        ),
+        port: source_wan_port,
       },
-      advice: true,
-      connection_type: ConnectionType::PUBLIC,
-      identifier: 1,
-      extra_bytes: String::from("42"),
+      advice: flags.bit7,
+      connection_type: ConnectionType::decode((flags.bit0, flags.bit1)),
+      identifier: identifier,
+      extra_bytes: extra_bytes,
     }
   }
 }
+
+
+#[cfg(test)]
+mod tests {
+  use super::*;
+
+  #[test]
+  fn integration_test_creation() {
+    let i = IntroductionRequestPayload {
+      destination_address: Address {
+        address: Ipv4Addr::new(127, 0, 0, 1),
+        port: 8000,
+      },
+      source_lan_address: Address {
+        address: Ipv4Addr::new(42, 42, 42, 42),
+        port: 8000,
+      },
+      source_wan_address: Address {
+        address: Ipv4Addr::new(255, 255, 255, 255),
```

Comment:
It may not matter now but it maybe consider not using broadcast addresses as they may be handled differently later in development

Reply:
good point. this is just a test of course but different test cases might be safer